python/lsst/pipe/base/graph/graph.py

```diff
@@ -614,19 +637,75 @@ def save(self, file):
         file : `io.BufferedIOBase`
             File to write pickle data open in binary mode.
         """
-        pickle.dump(self, file)
+        buffer = self._buildSaveObject()
+        file.write(buffer.getvalue())
+
+    def _buildSaveObject(self) -> io.BytesIO:
+        # make some containers
+        pickleData = []
+        nodeMap = {}
+        taskDefMap = {}
+        protocol = 3
+
+        # counter for the number of bytes processed thus far
+        count = 0
+        # serialize out the task Defs recording the start and end bytes of each
+        # taskDef
+        for taskDef in self.taskGraph:
+            # compressing has very little impact on saving or load time, but
+            # a large impact on on disk size, so it is worth doing
+            dump = lzma.compress(pickle.dumps(taskDef, protocol=protocol))
+            taskDefMap[taskDef.label] = (count, count+len(dump))
+            count += len(dump)
+            pickleData.append(dump)
+
+        # serialize the nodes, recording the start and end bytes of each node
+        self_list = list(self)
+        for node in self_list:
+            node = copy.copy(node)
+            taskDef = node.taskDef
+            object.__setattr__(node, 'taskDef', taskDef.label)
```

Comment:
This is because in normal operation, QuantumNodes are frozen dataclasses. However, because of the way things are being serialized for partial reads, there is no longer one big pickle file. Pickle is really efficient in that it would not serialize the same object multiple times, which we now were, and it was huge and slow. This is why the TaskDef objects were normalized out to their own save block.

We could have a different structure for this, but it seemed like a lot of overhead for something that is created to be dumped to disk, and then read back in and turned into something else. Especially as the structure of the class would be exactly the same, just with a different type declared. If you insist I will do it, but it seems to me to be one of those cases for which python flexibility is designed.

On a side note, I put in a comment around one of those blocks, but I guess I forgot to put it where it is saved as well. I will for sure add that.